### PR TITLE
[clang] Support __attribute__((ifunc(...))) on Darwin platforms

### PR DESCRIPTION
**Pull request by jroelofs**
Unlike ELF targets, MachO does not support the same kind of dynamic symbol
resolution at load time.  Instead, the corresponding MachO feature resolves
symbols lazily on first call.

### CODE REVIEW

**Comment by echristo** [comment]
Do you need an OS version check here?

**Comment by tahonermann** [comment]
Added @erichkeane as a reviewer.

I recommend extending test coverage to all of the following tests:
- `clang/test/CodeGenCXX/externc-ifunc-resolver.cpp`
- `clang/test/SemaCXX/externc-ifunc-resolver.cpp`
- `clang/test/CodeGen/ifunc.c` (already done)
- `clang/test/CodeGen/attr-ifunc.cpp`
- `clang/test/CodeGen/attr-ifunc.c`

**Comment by jroelofs** [comment]
This is part of a patch series:
    * #73686
    * #73688
    * #73687
    * #73685

**Comment by echristo** [comment]
On Wed, Nov 29, 2023 at 9:37 AM Jon Roelofs ***@***.***>
wrote:

> Do you need an OS version check here?
>
> I don't think so, at least not yet. With #73686
> <https://github.com/llvm/llvm-project/pull/73686>, support for this is
> entirely up to the compiler.
>
Aha I was wondering where the implementation side was :)

It looks like you don't necessarily need linker support as you're putting
it into the binary on occasion as well. This could work for now, but if you
plan on using this extensively it might be worth something else (as you
note).

File: clang/test/CodeGen/ifunc.c [hunk 1]
```diff
@@ -3,6 +3,10 @@
 // RUN: %clang_cc1 -triple i386-unknown-linux-gnu -fsanitize=thread -O2 -emit-llvm -o - %s | FileCheck %s --check-prefix=SAN
 // RUN: %clang_cc1 -triple i386-unknown-linux-gnu -fsanitize=address -O2 -emit-llvm -o - %s | FileCheck %s --check-prefix=SAN
 // RUN: %clang_cc1 -triple i386-unknown-linux-gnu -fsanitize=memory -O2 -emit-llvm -o - %s | FileCheck %s --check-prefix=SAN
+// RUN: %clang_cc1 -triple x86_64-apple-macosx -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-apple-macosx -O2 -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-apple-macosx -fsanitize=thread -O2 -emit-llvm -o - %s | FileCheck %s --check-prefix=MACSAN
+// RUN: %clang_cc1 -triple x86_64-apple-macosx -fsanitize=address -O2 -emit-llvm -o - %s | FileCheck %s --check-prefix=MACSAN
```
Comment (MaskRay) [inline clang/test/CodeGen/ifunc.c:9]: Consider changing one extra apple RUN line to arm64 to add some nuance.

File: clang/include/clang/Basic/AttrDocs.td [hunk 2]
```diff
@@ -5408,7 +5408,9 @@ considered inline.
 Not all targets support this attribute. ELF target support depends on both the
 linker and runtime linker, and is available in at least lld 4.0 and later,
 binutils 2.20.1 and later, glibc v2.11.1 and later, and FreeBSD 9.1 and later.
-Non-ELF targets currently do not support this attribute.
+MachO targets support it, but with slightly different semantics: the resolver is
```
Comment (MaskRay) [inline clang/include/clang/Basic/AttrDocs.td:5411]: In docs and comments, we prefer the canonical spelling: Mach-O

Comment (jroelofs) [reply clang/include/clang/Basic/AttrDocs.td:5411]: TIL, thanks!

**Review by MaskRay — APPROVED** [review]

**Review by ahmedbougacha — APPROVED** [review]

**Review by echristo — APPROVED** [review]
SGTM.

**Pull request merged**